File: daggercore/src/main/java/com/worldsnas/daggercore/navigator/changehandler/CircularRevealChangeHandlerCompat.java
package com.worldsnas.daggercore.navigator.changehandler;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.os.Build;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;

public class CircularRevealChangeHandlerCompat extends CircularRevealChangeHandler {

    public CircularRevealChangeHandlerCompat() { }

    public CircularRevealChangeHandlerCompat(@NonNull View fromView, @NonNull View containerView) {
        super(fromView, containerView);
    }

    @Override @NonNull
    protected Animator getAnimator(@NonNull ViewGroup container, View from, View to, boolean isPush, boolean toAddedToContainer) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return super.getAnimator(container, from, to, isPush, toAddedToContainer);
        } else {
            AnimatorSet animator = new AnimatorSet();
            if (to != null) {
                float start = toAddedToContainer ? 0 : to.getAlpha();
                animator.play(ObjectAnimator.ofFloat(to, View.ALPHA, start, 1));
            }

            if (from != null) {
                animator.play(ObjectAnimator.ofFloat(from, View.ALPHA, 0));
            }

            return animator;
        }
    }
}

File: navigation/src/androidMain/kotlin/com/worldsnas/navigation/changehandler/CircularRevealChangeHandlerCompat.kt
package com.worldsnas.navigation.changehandler

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.os.Build
import android.view.View
import android.view.ViewGroup
import androidx.annotation.Keep

class CircularRevealChangeHandlerCompat : CircularRevealChangeHandler {
    @Keep
    constructor() {
    }

    constructor(fromView: View, containerView: View) : super(
        fromView,
        containerView
    ) {
    }

    constructor(cx: Int, cy: Int) : super(cx, cy) {}
    constructor(cx: Int, cy: Int, duration: Long) : super(cx, cy, duration) {}

    override fun getAnimator(
        container: ViewGroup,
        from: View?,
        to: View?,
        isPush: Boolean,
        toAddedToContainer: Boolean
    ): Animator {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            super.getAnimator(container, from, to, isPush, toAddedToContainer)
        } else {
            val animator = AnimatorSet()
            if (to != null) {
                val start: Float = if (toAddedToContainer) 0F else to.alpha
                animator.play(
                    ObjectAnimator.ofFloat(
                        to,
                        View.ALPHA,
                        start,
                        1f
                    )
                )
            }
            if (from != null) {
                animator.play(
                    ObjectAnimator.ofFloat(
                        from,
                        View.ALPHA,
                        0f
                    )
                )
            }
            animator
        }
    }
}